File: src/main/java/tn/esprit/entities/Participation.java
package tn.esprit.entities;

import java.util.Date;

/**
 * Entité alignée sur la table `participation`.
 *
 * Colonnes DB :
 *   id, nom, prenom, email, telephone, date_inscription,
 *   code_qr, statut, position_attente, evenement_id
 */
public class Participation {

    private int     id;
    private String  nom;
    private String  prenom;
    private String  email;
    private String  telephone;
    private Date    dateInscription;
    private String  codeQr;
    private String  statut;
    private Integer positionAttente;
    private int     evenementId;

    /** Champ de jointure (non persisté) – titre affiché dans les tableaux */
    private String nomEvenement;

    // ── Constructeurs ──────────────────────────────────────────────────────────

    public Participation() {}

    // ── Getters / Setters ──────────────────────────────────────────────────────

    public int getId()              { return id; }
    public void setId(int id)       { this.id = id; }

    /** Alias de compatibilité */
    public int getId_participation()            { return id; }
    public void setId_participation(int id)     { this.id = id; }

    public String getNom()              { return nom; }
    public void setNom(String nom)      { this.nom = nom; }

    public String getPrenom()               { return prenom; }
    public void setPrenom(String prenom)    { this.prenom = prenom; }

    /** Renvoie "nom prénom" – alias de compatibilité */
    public String getNom_participation() {
        return (nom != null ? nom : "") + " " + (prenom != null ? prenom : "");
    }
    public void setNom_participation(String v) { this.nom = v; }

    public String getEmail()                { return email; }
    public void setEmail(String email)      { this.email = email; }

    public String getTelephone()                    { return telephone; }
    public void setTelephone(String telephone)      { this.telephone = telephone; }

    public Date getDateInscription()                        { return dateInscription; }
    public void setDateInscription(Date dateInscription)    { this.dateInscription = dateInscription; }

    public String getCodeQr()               { return codeQr; }
    public void setCodeQr(String codeQr)    { this.codeQr = codeQr; }

    public String getStatut()               { return statut; }
    public void setStatut(String statut)    { this.statut = statut; }

    /** Alias de compatibilité */
    public String getStatutParticipation()          { return statut; }
    public void setStatutParticipation(String v)    { this.statut = v; }

    /** Colonne position_attente (nullable) */
    public Integer getPositionAttente()                         { return positionAttente; }
    public void setPositionAttente(Integer positionAttente)     { this.positionAttente = positionAttente; }

    public int getEvenementId()                 { return evenementId; }
    public void setEvenementId(int evenementId) { this.evenementId = evenementId; }

    /** Alias de compatibilité avec l'ancien code getId_evenemnt() */
    public int getId_evenemnt()             { return evenementId; }
    public void setId_evenemnt(int v)       { this.evenementId = v; }

    public String getNomEvenement()                     { return nomEvenement; }
    public void setNomEvenement(String nomEvenement)    { this.nomEvenement = nomEvenement; }

    /** Alias de compatibilité */
    public String getNom_evenement()        { return nomEvenement; }
    public void setNom_evenement(String v)  { this.nomEvenement = v; }

    public boolean isEnAttente() { return "en_attente".equals(statut); }

    // ── Alias de compatibilité – anciens champs supprimés ─────────────────────

    /** Ancien champ utilisateur_id supprimé – conservé pour éviter erreurs de compilation */
    public void setUtilisateur_id(int v)    { /* ignoré */ }
    public int  getUtilisateur_id()         { return 0; }

    /** Ancien champ feedback supprimé – conservé pour éviter erreurs de compilation */
    public void setFeedback(String v)       { /* ignoré */ }
    public String getFeedback()             { return ""; }

    @Override
    public String toString() {
        return "Participation{id=" + id + ", nom='" + nom + " " + prenom
                + "', statut='" + statut + "'}";
    }
}

File: src/Entity/Participation.php
<?php

namespace App\Entity;

use App\Repository\ParticipationRepository;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: ParticipationRepository::class)]
class Participation
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    /** @phpstan-ignore property.unusedType */
    private ?int $id = null;

    #[ORM\Column(length: 100)]
    private string $nom = '';

    #[ORM\Column(length: 100)]
    private string $prenom = '';

    #[ORM\Embedded(class: Email::class, columnPrefix: false)]
    private Email $email;

    #[ORM\Column(length: 20, nullable: true)]
    private ?string $telephone = null;

    #[ORM\Column(type: 'datetime_immutable')]
    private \DateTimeImmutable $dateInscription;

    #[ORM\Column(length: 255, nullable: true)]
    private ?string $codeQr = null;

    #[ORM\Column(length: 30)]
    private string $statut = '';

    #[ORM\Column(nullable: true)]
    private ?int $positionAttente = null;

    #[ORM\ManyToOne(inversedBy: 'participations')]
    #[ORM\JoinColumn(nullable: false, onDelete: 'CASCADE')]
    private ?Evenement $evenement = null;

    public function __construct()
    {
        $this->email = $this->createEmptyEmail();
        $this->dateInscription = new \DateTimeImmutable();
    }

    private function createEmptyEmail(): Email
    {
        $reflection = new \ReflectionClass(Email::class);
        /** @var Email $email */
        $email = $reflection->newInstanceWithoutConstructor();
        $prop  = $reflection->getProperty('value');
        $prop->setAccessible(true);
        $prop->setValue($email, '');
        return $email;
    }

    public function getId(): ?int { return $this->id; }

    public function getNom(): string { return $this->nom; }
    public function setNom(string $nom): static { $this->nom = $nom; return $this; }

    public function getPrenom(): string { return $this->prenom; }
    public function setPrenom(string $prenom): static { $this->prenom = $prenom; return $this; }

    public function getEmail(): Email { return $this->email; }
    public function setEmail(Email $email): static { $this->email = $email; return $this; }

    public function getTelephone(): ?string { return $this->telephone; }
    public function setTelephone(?string $telephone): static { $this->telephone = $telephone; return $this; }

    public function getDateInscription(): \DateTimeImmutable { return $this->dateInscription; }
    protected function setDateInscription(\DateTimeImmutable $d): static { $this->dateInscription = $d; return $this; }

    public function getCodeQr(): ?string { return $this->codeQr; }
    public function setCodeQr(?string $codeQr): static { $this->codeQr = $codeQr; return $this; }

    public function getStatut(): string { return $this->statut; }
    public function setStatut(string $statut): static { $this->statut = $statut; return $this; }

    public function getPositionAttente(): ?int { return $this->positionAttente; }
    public function setPositionAttente(?int $p): static { $this->positionAttente = $p; return $this; }

    public function isEnAttente(): bool { return $this->statut === 'en_attente'; }

    public function getEvenement(): ?Evenement { return $this->evenement; }
    public function setEvenement(?Evenement $evenement): static { $this->evenement = $evenement; return $this; }
}